plugins/modules/net_tools/dnsimple.py

```diff
@@ -201,135 +367,116 @@ def main():
-                        'type': record_type,
-                        'content': value,
-                    }
-                    if ttl:
-                        data['ttl'] = ttl
-                    if priority:
-                        data['priority'] = priority
                     if module.check_mode:
                         module.exit_json(changed=True)
                     else:
-                        module.exit_json(changed=True, result=client.add_record(str(domain), data)['record'])
-
+                        response = create_record(client, account, domain, record,
+                                                 record_type, value, ttl, priority)
+                        module.exit_json(changed=True, result=response)
             # state is absent
             else:
                 if rr:
                     if not module.check_mode:
-                        client.delete_record(str(domain), rr['id'])
+                        delete_record(client, account, domain, rr['id'])
                     module.exit_json(changed=True)
                 else:
                     module.exit_json(changed=False)
 
         # Make sure these record_ids either all exist or none
         if domain and record_ids:
-            current_records = [str(r['record']['id']) for r in client.records(str(domain))]
-            wanted_records = [str(r) for r in record_ids]
+            current_records = get_records(client, account, domain, dnsimple_filter=None)
+            current_record_ids = [str(d['id']) for d in current_records]
+            wanted_record_ids = [str(r) for r in record_ids]
             if state == 'present':
-                difference = list(set(wanted_records) - set(current_records))
+                difference = list(set(wanted_record_ids) - set(current_record_ids))
                 if difference:
                     module.fail_json(msg="Missing the following records: %s" % difference)
                 else:
                     module.exit_json(changed=False)
-
             # state is absent
             else:
-                difference = list(set(wanted_records) & set(current_records))
+                difference = list(set(wanted_record_ids) & set(current_record_ids))
                 if difference:
                     if not module.check_mode:
                         for rid in difference:
-                            client.delete_record(str(domain), rid)
+                            delete_record(client, account, domain, rid)
                     module.exit_json(changed=True)
                 else:
                     module.exit_json(changed=False)
 
     except DNSimpleException as e:
-        module.fail_json(msg="Unable to contact DNSimple: %s" % e.message)
-
+        if DNSIMPLE_MAJOR_VERSION > 1:
+            module.fail_json(msg="DNSimple exception: %s" % e.message)
+        else:
+            module.fail_json(msg="DNSimple exception: %s" % str(e.args[0]['message']))
```

Comment:
Why not using `e.message` here?

Reply:
As far as I could see, `python-dnsimple` < 2 did not implement the `message` instance variable in the `DNSimpleException` class (see [here](https://github.com/dnsimple/dnsimple-python/blob/1.2.0/dnsimple/dnsimple.py#L29)). If I understand it correctly, the [current implementation](https://github.com/ansible-collections/community.general/blob/main/plugins/modules/net_tools/dnsimple.py#L331) of the exception handling in this moduel does not work for `python-dnsimple` < 2. In case a `DNSimpleException` is raised, python raises another exception complaining the 'DNSimpleException' object has no attribute 'message'. I created a [gist](https://gist.github.com/rekup/8a28b02411d7f0ca84ac3b73fd840993) with the full error message.
The `DNSimpleException` class in `python-dnsimple` >=2 however does implement this: https://github.com/dnsimple/dnsimple-python/blob/main/dnsimple/exceptions.py#L3

Reply:
Interesting, good catch :)